### ISSUE TITLE
`mkMessage` eats 30 GB of memory during compile

### ISSUE THREAD

**Issue by ygale** [yesodweb/yesod#1487]
On Ubuntu 16.04, LTS-9.17, when compiling `mkMessage`, GHC spikes to about 30 GB of memory usage.

This site has 37 message files in the `messages` folder. There are 540 messages in `en.msg`, of which 22 have one or more typed parameters. In each of the other 36 message files, there are 485 messages, of which 20 have one or more typed parameters.

We expect to need to compile sites with many more message files than this in the future. So this bug is worrying.

Here is the complete contents of the module in which `mkMessage` is compiled, where the memory spike occurs:

    module Foundation.Messages
    
    where
    
    import Prelude
    import Data.Text
    import Yesod hiding (PageContent)
    
    import Foundation.App
    
    -- Set up i18n messages. See the message folder.
    mkMessage "App" "messages" "en"

**Comment by snoyberg** [comment]
I'm worried that this might be inherent to large sum types in GHC.

Any chance on providing a reproducing case for this? I have a strong feeling that ultimately this will end up as a GHC bug report.

**Comment by ygale** [comment]
My feeling too. I'll see if I can come up with something.

**Comment by ygale** [comment]
https://github.com/ygale/yesod-bug1487

on LTS-10.5

**Comment by ygale** [comment]
Wow it actually ate over 50 GB.

**Comment by ncaq** [comment]
What is the optimization flag?
I had consumed a lot of memory when I installed a large amount of files below static, but when I put optimization in the default state, the problem disappeared.

**Comment by ygale** [comment]
I tested using`yesod keter`  with no special parameters.

**Comment by sbditto85** [comment]
I had problems in the LTS-10.x compiling models where it would eat 30gb with a stack build but work with stack build —fast. Ended up switching to LTS-9.21 and it worked. Guessing 8.2.2 has a bug somewhere (there is a known issue on I think 8.2.1, but on mobile so idk what the link is). If you need a link to that project let me know.

**Comment by ygale** [comment]
@sbditto85 The reproduction repo is on LTS-10.5, but the problem occurs in our application using LTS-9.17, as noted above. You can test the reproduction repo on 9.17 or 9.21, I think you'll see the issue there too.

I think you're right that it's a GHC bug. It might not even be a recent regression, either.

**Comment by ncaq** [comment]
> I tested usingyesod keter with no special parameters. On Ubuntu 16.04 with LTS-10.5, as noted above.

Can you paste `package.yaml` `ghc-options:`?

**Comment by ygale** [comment]
@ncaq everything needed is in the reproduction repo I linked to above. The `package.yaml` is there, and `ghc-options:` isn't explicitly set anywhere (unless the scaffolding does so, in which case you'll find it in the repo).

Also - to clarify - `yesod keter` is what we use in real life, when the bug happened to us. But for my reproduction, I just ran `stack build` and it reproduced.

I tried to get a reproduction without TH, and was unable to do so. Here is a [gist of a module](https://gist.github.com/92347aa93d226e31f977a0b62b443aa7) that simulates what yesod does in `mkMessage` for my reproduction, but without TH. The module in that gist compiles just fine in GHC 8.2.2.

So it looks like a TH bug. The gory details of the TH in `mkMessage` is beyond my expertise, so I will need some help boiling this down to a reproduction that can be reported in a GHC Trac issue. Thanks!

**Comment by ygale** [comment]
Is there anything I can do to help move this issue forward?

**Comment by snoyberg** [comment]
To clarify what you mean with move this forward: you're asking for help in reproducing the Gist above to a code sample that doesn't use Yesod, so that it can be reported to GHC HQ?

**Comment by ygale** [comment]
@snoyberg right. In fact, I can't even report it truthfully as a GHC bug on my own, becasue I don't understand the `mokMessage` TH code well enough to eliminate the possiblity that it's a yesod bug.

If you feel we're already ready to report a GHC bug, and that my yesod-based repro linked above is enough, I'd be happy to do that. Otherwise - what needs to be done to move forward?

Note that the gist is *not* a repro. It is a *non-repro* to demonstrate that TH is required to reproduce. For convenience, here is a [repeat of the link to the repro](https://github.com/ygale/yesod-bug1487).

**Comment by snoyberg** [comment]
I've just opened a PR with a minimized repro. It now only depends on packages that ship with GHC.

**Comment by ygale** [comment]
@snoyberg Thanks! It's missing the `I18N` module though. Could you please add that? Thanks!

**Comment by ygale** [comment]
Apart from the GHC bug, wouldn't it make sense for `mkMessage` to create a singleton for each message rather than a massive monolithic type? Semantically, it's hard for me to imagine a use case where the type `AppMessage` (one of the specific messages defined in the `message` folder) makes more sense than `forall r. RenderMessage r => r` (something that can be localized). Although technically I suppose the extra polymorphism could be more awkward in some cases.

**Comment by snoyberg** [comment]
I have no plans to make such a change to the library. If someone else wants to experiment with it and see if it works out well, I'd be happy to review, but I haven't thought through the details at all myself.
